util/src/main/java/io/kubernetes/client/util/credentials/AuthenticationRefresher.java

```diff
@@ -0,0 +1,96 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package io.kubernetes.client.util.credentials;
+
+import java.lang.ref.WeakReference;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import io.kubernetes.client.openapi.ApiClient;
+
+/**
+ * Wraps an existing {@link Authentication} and refreshes it every
+ * expirationSeconds.
+ *
+ * Can be used with ClientBuilder as such:
+ *
+ * <pre> ClientBuilder.standard() .withAuthentication(new
+ * AuthenticationRefresher(new KubeconfigAuthentication(), 60)) .build(); </pre>
+ *
+ * The AuthenticationRefresher maintains a {@link WeakReference} to the
+ * ApiClient passed into provide. When the client instance goes out of scope,
+ * the refresh timer will be cancelled and released.
+ */
+public class AuthenticationRefresher implements Authentication {
+
+    private static final Logger log = LoggerFactory.getLogger(AuthenticationRefresher.class);
+
+    private final Authentication delegateAuthentication;
+    private final Long expirationSeconds;
+
+    private final ScheduledExecutorService executor;
+    private ScheduledFuture<?> currentSchedule;
+
+    public AuthenticationRefresher(Authentication delegateAuthentication, long expirationSeconds) {
+        this.delegateAuthentication = delegateAuthentication;
+        this.expirationSeconds = expirationSeconds;
+        this.executor = Executors.newSingleThreadScheduledExecutor();
+        log.debug("AuthenticationRefresher initialized with expirationSeconds: " + expirationSeconds);
+    }
+
+    /**
+     * Calls delegateAuthentication every expirationSeconds
+     */
+    @Override
+    public void provide(ApiClient client) {
+
+        synchronized (this) {
```

Comment:
Why do this lazily on the first call to provide vs. doing it in the contructor? Do you expect that the ApiClient will get garbage collected repeatedly but this authentication will stick around?

Reply:
Still wondering about this. Why not just create the executor once in the constructor and be done with it? This current logic seems convulted and I'm not sure I see the value of the lazy evaluation.